### PR TITLE
Fix offset issue of bitcast and concat.

### PR DESCRIPTION
**Pull request by zhangyang2057**
- [x] Add bfs_ir_pre_order_visitor.
- [x] Fix bitcast offset  and add reproduced test case.
- [x] Use offset instead of begin shape to avoid shape rank mismatch 
- [x] Move concat from scheduler to pass

### CODE REVIEW

File: src/transforms/neutral/optimize_allocation.cpp [hunk 1]
```diff
@@ -200,7 +200,7 @@ void alias_bitcast_buffer_pass::run_core(graph &graph, [[maybe_unused]] nncase::
 
                     // owner transfered to output
                     shape_t begin(b->output().shape().size(), 0);
-                    in_buf.parent() = { &out_buf, begin, b->output().shape() };
+                    in_buf.parent() = { &out_buf, out_buf.parent() ? out_buf.parent()->begin : begin, b->output().shape() };
```
Comment (sunnycase) [inline src/transforms/neutral/optimize_allocation.cpp:203]: `begin` here is local which means relative to it's parent (bitcast's output buffer). The global begin should be calculated at parent broadcast stage (currently at `fix_concat_indices`).

File: src/schedule/scheduler.cpp [hunk 2]
```diff
@@ -495,9 +495,9 @@ schedule_result scheduler::schedule(bool skip_buffer_alias)
         context.target = &target_;
 
         context.make_logical_buffers();
+        context.fix_concat_indices();
```
Comment (sunnycase) [inline src/schedule/scheduler.cpp:498]: The order shouldn't be exchanged.

**Review by sunnycase — APPROVED** [review]

File: src/ir/visitor.cpp [hunk 3]
```diff
@@ -91,3 +92,29 @@ bool dfs_ir_post_order_visitor::visit_strategy(node &node)
...
+            {
+                auto &in_node = in->connection()->owner();
+                if (!visited(in_node))
+                    nodes.push(&in_node);
+            }
+        }
+    }
+
+    return false;
+}
```
Comment (zhangyang2057) [inline src/ir/visitor.cpp:120]: new line

File: include/nncase/ir/visitor.h [hunk 4]
```diff
@@ -53,6 +53,14 @@ class NNCASE_API dfs_ir_post_order_visitor : public ir_visitor
 private:
 };
 
+class NNCASE_API bfs_ir_visitor : public ir_visitor
```
Comment (zhangyang2057) [inline include/nncase/ir/visitor.h:56]: rename as bfs_ir_pre_order_visitor

**Pull request merged**